backends/gstreamer/src/source.rs

```diff
@@ -38,113 +28,180 @@ mod imp {
         )
     }
 
+    // The actual data structure that stores our values. This is not accessible
+    // directly from the outside.
     pub struct ServoSrc {
+        cat: gst::DebugCategory,
         appsrc: gst_app::AppSrc,
+        srcpad: gst::GhostPad,
     }
 
     impl ServoSrc {
-        fn init(_bin: &Bin) -> Box<BinImpl<Bin>> {
-            let appsrc = gst::ElementFactory::make("appsrc", None)
-                .map(|elem| elem.downcast::<AppSrc>().unwrap())
-                .expect("Could not create appsrc element");
+        pub fn set_size(&self, size: i64) {
+            if self.appsrc.get_size() == -1 {
+                self.appsrc.set_size(size);
+            }
+        }
 
-            appsrc.set_max_bytes(MAX_SRC_QUEUE_SIZE);
-            appsrc.set_property_block(false);
-            appsrc.set_property_format(gst::Format::Bytes);
+        inner_appsrc_proxy!(end_of_stream, Result<gst::FlowSuccess, gst::FlowError>);
+        inner_appsrc_proxy!(get_current_level_bytes, u64);
+        inner_appsrc_proxy!(get_max_bytes, u64);
+        inner_appsrc_proxy!(push_buffer, buffer, gst::Buffer, Result<gst::FlowSuccess, gst::FlowError>);
+        inner_appsrc_proxy!(set_callbacks, callbacks, gst_app::AppSrcCallbacks, ());
+        inner_appsrc_proxy!(set_stream_type, type_, gst_app::AppStreamType, ());
 
-            // At this point the bin is not completely created,
-            // so we cannot add anything to it yet.
-            // We have to wait until ObjectImpl::constructed.
-            Box::new(Self { appsrc })
+        fn query(&self, pad: &gst::GhostPad, query: &mut gst::QueryRef) -> bool {
+            gst_log!(self.cat, obj: pad, "Handling query {:?}", query);
+
+            // In order to make buffering/downloading work as we want, apart from
+            // setting the appropriate flags on the player playbin,
+            // the source needs to either:
+            //
+            // 1. be an http, mms, etc. scheme
+            // 2. report that it is "bandwidth limited".
+            //
+            // 1. is not straightforward because we are using a servosrc scheme for now.
+            // This may change in the future if we end up handling http/https/data
+            // URIs, which is what WebKit does.
+            //
+            // For 2. we need to make servosrc handle the scheduling properties query
+            // to report that it "is bandwidth limited".
+            let ret = match query.view_mut() {
+                gst::QueryView::Scheduling(ref mut q) => {
+                    let flags =
+                        gst::SchedulingFlags::SEQUENTIAL | gst::SchedulingFlags::BANDWIDTH_LIMITED;
+                    q.set(flags, 1, -1, 0);
+                    q.add_scheduling_modes(&[gst::PadMode::Push]);
+                    true
+                }
+                _ => {
+                    if let Some(target) = pad.get_target() {
+                        target.query(query)
```

Comment:
Why not `pad.query_default(Some(element), query)`? This is not the same and you probably want the default behaviour instead of just proxying everything.

Reply:
Oops! sorry! You already explained that to me and I stubbornly insist :) fixing it